### PR TITLE
Add mask2former fp16 support

### PR DESCRIPTION
**Pull request by pedrohml**
# What does this PR do?

Add float16 support to mask2former module and derivatives
Some operations and conversions were fixed in order to propagate the chosen user _dtype_ (ex: float32, float16) for input and embeddings. In this way, we can pass in fp16 tensors for inference.

## Who can review?

Similar PRs were reviewed by: @alaradirik, @amyeroberts, @LysandreJik

### CODE REVIEW

**Comment by amyeroberts** [comment]
Hi @pedrohml, thanks for opening this PR! 

Could you add some tests for these models e.g. [like this one for ViT](https://github.com/huggingface/transformers/blob/dd9d45b6ecc0861847e21d461187711331a56138/tests/models/vit/test_modeling_vit.py#L320)

**Comment by pedrohml** [comment]
@amyeroberts thanks for review. do you mind to double check new changes ?

**Comment by amyeroberts** [comment]
@pedrohml Thanks for iterating and adding some tests. The tests added should match the pattern of the ones [I linked to](https://github.com/huggingface/transformers/blob/dd9d45b6ecc0861847e21d461187711331a56138/tests/models/vit/test_modeling_vit.py#L320), rather than call integration tests that check logit values. There are two reasons for this: 
*  An independent test enables us to select as small as possible checkpoint / config to keep the CI runs as fast as possible
* It might not be robust, as we can expect some differences in the output logits. Out of interest, did you run the slow tests to confirm these pass?

**Comment by pedrohml** [comment]
> @pedrohml Thanks for iterating and adding some tests. The tests added should match the pattern of the ones [I linked to](https://github.com/huggingface/transformers/blob/dd9d45b6ecc0861847e21d461187711331a56138/tests/models/vit/test_modeling_vit.py#L320), rather than call integration tests that check logit values. There are two reasons for this:
> 
> * An independent test enables us to select as small as possible checkpoint / config to keep the CI runs as fast as possible
> * It might not be robust, as we can expect some differences in the output logits. Out of interest, did you run the slow tests to confirm these pass?

@amyeroberts 
Thanks for the heads up.. I managed to simplify the fp16 tests. Also, I was able to run tests locally using gpu to confirm all tests go green. This helped me to fix a mistake for oneformer workaround.
I appreciate if you can review again. Feel free to add something else.

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this functionality and iterating!

File: src/transformers/models/oneformer/modeling_oneformer.py [hunk 1]
```diff
@@ -1380,7 +1380,7 @@ def forward(
         sources = []
         position_embeddings_list = []
         for level, source in enumerate(features[::-1][: self.num_feature_levels]):
-            feats = source.float()
+            feats = source.to(source.dtype)
```
[the remark this answers is not in the document]
Comment (pedrohml) [reply src/transformers/models/oneformer/modeling_oneformer.py:1383]: I made this noop when replicating the changes across all \*formers. Will remove. Good catch!
